File: KKY_Tool_Revit_2019-2023/UI/Hub/UiBridgeExternalEvent.ParamProp.vb
Option Explicit On
Option Strict On

Imports System
Imports System.Linq
Imports System.Data
Imports Autodesk.Revit.UI
Imports KKY_Tool_Revit.Services

Namespace UI.Hub

    Partial Public Class UiBridgeExternalEvent

        Private Shared _lastParamResult As ParamPropagateService.SharedParamRunResult

        ' === sharedparam:list ===
        Private Sub HandleSharedParamList(app As UIApplication, payload As Object)
            Try
                Dim status = SharedParameterStatusService.GetStatus(app)
                If status Is Nothing OrElse Not String.Equals(status.Status, "ok", StringComparison.OrdinalIgnoreCase) Then
                    Dim msg = If(String.IsNullOrWhiteSpace(status?.WarningMessage), "Shared Parameter 파일 상태가 올바르지 않습니다.", status.WarningMessage)
                    SendToWeb("sharedparam:list", New With {.ok = False, .message = msg, .items = New List(Of Object)()})
                    Return
                End If

                Dim res = ParamPropagateService.GetSharedParameterDefinitions(app)
                Dim items = SharedParameterStatusService.ListDefinitions(app).Select(Function(d) New With {
                    .name = d.Name,
                    .guid = d.Guid,
                    .groupName = d.GroupName,
                    .dataTypeToken = d.DataTypeToken
                }).ToList()

                Dim shaped As Object = New With {
                    .ok = res IsNot Nothing AndAlso res.Ok,
                    .message = If(res Is Nothing, Nothing, res.Message),
                    .items = items,
                    .definitions = items.Select(Function(d) New With {
                        .groupName = d.groupName,
                        .name = d.name,
                        .paramType = d.dataTypeToken,
                        .visible = True
                    }).ToList(),
                    .targetGroups = If(res?.TargetGroups, New List(Of ParamPropagateService.ParameterGroupOption)()).Select(Function(g) New With {
                        .id = g.Id,
                        .name = g.Name
                    }).ToList()
                }
                SendToWeb("sharedparam:list", shaped)
            Catch ex As Exception
                SendToWeb("sharedparam:list", New With {.ok = False, .message = ex.Message, .items = New List(Of Object)()})
                SendToWeb("revit:error", New With {.message = ex.Message})
            End Try
        End Sub

        ' === sharedparam:status ===
        Private Sub HandleSharedParamStatus(app As UIApplication, payload As Object)
            Try
                Dim status = SharedParameterStatusService.GetStatus(app)
                Dim shaped As Object = New With {
                    .path = status.Path,
                    .isSet = status.IsSet,
                    .existsOnDisk = status.ExistsOnDisk,
                    .canOpen = status.CanOpen,
                    .status = status.Status,
                    .statusLabel = status.StatusLabel,
                    .warning = status.WarningMessage,
                    .errorMessage = status.ErrorMessage
                }
                SendToWeb("sharedparam:status", shaped)
            Catch ex As Exception
                SendToWeb("sharedparam:status", New With {
                    .status = "error",
                    .statusLabel = "조회 실패",
                    .warning = "Shared Parameter 상태 조회에 실패했습니다.",
                    .errorMessage = ex.Message
                })
            End Try
        End Sub

        ' === sharedparam:run / paramprop:run ===
        Private Sub HandleSharedParamRun(app As UIApplication, payload As Object)
            Try
                Dim sharedStatus = SharedParameterStatusService.GetStatus(app)
                If sharedStatus Is Nothing OrElse Not String.Equals(sharedStatus.Status, "ok", StringComparison.OrdinalIgnoreCase) Then
                    Dim msg = If(String.IsNullOrWhiteSpace(sharedStatus?.WarningMessage), "Shared Parameter 파일 상태가 올바르지 않습니다.", sharedStatus.WarningMessage)
                    SendToWeb("sharedparam:done", New With {.ok = False, .status = "blocked", .message = msg})
                    SendToWeb("paramprop:done", New With {.ok = False, .status = "blocked", .message = msg})
                    SendToWeb("revit:error", New With {.message = "공유 파라미터 연동 실패: " & msg})
                    SendToWeb("sharedparam:status", New With {
                        .path = sharedStatus?.Path,
                        .isSet = sharedStatus?.IsSet,
                        .existsOnDisk = sharedStatus?.ExistsOnDisk,
                        .canOpen = sharedStatus?.CanOpen,
                        .status = sharedStatus?.Status,
                        .statusLabel = sharedStatus?.StatusLabel,
                        .warning = sharedStatus?.WarningMessage,
                        .errorMessage = sharedStatus?.ErrorMessage
                    })
                    Return
                End If

                Dim req As ParamPropagateService.SharedParamRunRequest = ParamPropagateService.SharedParamRunRequest.FromPayload(payload)
                Dim res = ParamPropagateService.Run(app, req, AddressOf ReportParamPropProgress)
                _lastParamResult = res

                Dim status = If(res Is Nothing, ParamPropagateService.RunStatus.Failed, res.Status)
                Dim ok As Boolean = (status = ParamPropagateService.RunStatus.Succeeded)

                Dim responsePayload As New Dictionary(Of String, Object)()
                responsePayload("ok") = ok
                responsePayload("status") = status.ToString().ToLowerInvariant()
                responsePayload("message") = If(res IsNot Nothing, res.Message, Nothing)
                If res IsNot Nothing Then
                    Dim filteredDetails As List(Of ParamPropagateService.SharedParamDetailRow) = FilterParamPropIssueDetails(res.Details)
                    responsePayload("report") = res.Report
                    responsePayload("details") = filteredDetails.Select(Function(d) New With {
                        .kind = d.Kind,
                        .family = d.Family,
                        .detail = d.Detail
                    }).ToList()
                End If

                SendToWeb("sharedparam:done", responsePayload)
                SendToWeb("paramprop:done", responsePayload)

                If Not ok Then
                    Dim msg As String = If(res Is Nothing, "실패", res.Message)
                    SendToWeb("revit:error", New With {.message = "공유 파라미터 연동 실패: " & msg})
                End If

            Catch ex As Exception
                SendToWeb("sharedparam:done", New With {.ok = False, .status = "failed", .message = ex.Message})
                SendToWeb("paramprop:done", New With {.ok = False, .status = "failed", .message = ex.Message})
                SendToWeb("revit:error", New With {.message = "공유 파라미터 연동 실패: " & ex.Message})
            End Try
        End Sub

        Private Function FilterParamPropIssueDetails(details As List(Of ParamPropagateService.SharedParamDetailRow)) As List(Of ParamPropagateService.SharedParamDetailRow)
            Dim src As List(Of ParamPropagateService.SharedParamDetailRow) = If(details, New List(Of ParamPropagateService.SharedParamDetailRow)())

            Dim dt As New DataTable("ParamPropUi")
            dt.Columns.Add("Type")
            dt.Columns.Add("Family")
            dt.Columns.Add("Detail")

            For Each d In src
                Dim row = dt.NewRow()
                row("Type") = If(d Is Nothing, "", If(d.Kind, ""))
                row("Family") = If(d Is Nothing, "", If(d.Family, ""))
                row("Detail") = If(d Is Nothing, "", If(d.Detail, ""))
                dt.Rows.Add(row)
            Next

            Dim filtered As DataTable = FilterIssueRowsCopy("paramprop", dt)
            Dim result As New List(Of ParamPropagateService.SharedParamDetailRow)()
            If filtered Is Nothing Then Return result

            For Each r As DataRow In filtered.Rows
                result.Add(New ParamPropagateService.SharedParamDetailRow With {
                    .Kind = Convert.ToString(r("Type")),
                    .Family = Convert.ToString(r("Family")),
                    .Detail = Convert.ToString(r("Detail"))
                })
            Next

            Return result
        End Function

        Private Sub ReportParamPropProgress(phase As String,
                                            phaseProgress As Double,
                                            current As Integer,
                                            total As Integer,
                                            message As String,
                                            target As String)
            SendToWeb("paramprop:progress", New With {
                .phase = phase,
                .phaseProgress = phaseProgress,
                .current = current,
                .total = total,
                .message = message,
                .target = target
            })
        End Sub

        ' === sharedparam:export-excel ===
        Private Sub HandleSharedParamExport(app As UIApplication, payload As Object)
            Try
                If _lastParamResult Is Nothing Then
                    SendToWeb("sharedparam:exported", New With {.ok = False, .message = "최근 실행 결과가 없습니다."})
                    Return
                End If

                Dim doAutoFit As Boolean = ParseExcelMode(payload)
                Dim saved As String = ParamPropagateService.ExportResultToExcel(_lastParamResult, doAutoFit)
                If String.IsNullOrWhiteSpace(saved) Then
                    SendToWeb("sharedparam:exported", New With {.ok = False, .message = "엑셀 내보내기가 취소되었습니다."})
                    Return
                End If

                SendToWeb("sharedparam:exported", New With {.ok = True, .path = saved})
            Catch ex As Exception
                SendToWeb("sharedparam:exported", New With {.ok = False, .message = ex.Message})
                SendToWeb("revit:error", New With {.message = "엑셀 내보내기 실패: " & ex.Message})
            End Try
        End Sub

    End Class

End Namespace

File: KKY_Tool_Revit_2019-2023/UI/Hub/UiBridgeExternalEvent.ParamProp.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using Autodesk.Revit.UI;
using KKY_Tool_Revit.Services;

namespace KKY_Tool_Revit.UI.Hub
{
    public sealed partial class UiBridgeExternalEvent
    {
        private static ParamPropagateService.SharedParamRunResult _lastParamResult;

        private void HandleSharedParamList(UIApplication app, object payload)
        {
            try
            {
                var status = SharedParameterStatusService.GetStatus(app);
                if (status == null || !string.Equals(status.Status, "ok", StringComparison.OrdinalIgnoreCase))
                {
                    var msg = string.IsNullOrWhiteSpace(status?.WarningMessage)
                        ? "Shared Parameter 파일 상태가 올바르지 않습니다."
                        : status.WarningMessage;
                    SendToWeb("sharedparam:list", new { ok = false, message = msg, items = new List<object>() });
                    return;
                }

                var res = ParamPropagateService.GetSharedParameterDefinitions(app);
                var items = SharedParameterStatusService.ListDefinitions(app).Select(d => new
                {
                    name = d.Name,
                    guid = d.Guid,
                    groupName = d.GroupName,
                    dataTypeToken = d.DataTypeToken
                }).ToList();

                var shaped = new
                {
                    ok = res != null && res.Ok,
                    message = res == null ? null : res.Message,
                    items,
                    definitions = items.Select(d => new
                    {
                        groupName = d.groupName,
                        name = d.name,
                        paramType = d.dataTypeToken,
                        visible = true
                    }).ToList(),
                    targetGroups = (res?.TargetGroups ?? new List<ParamPropagateService.ParameterGroupOption>()).Select(g => new
                    {
                        id = g.Id,
                        name = g.Name
                    }).ToList()
                };
                SendToWeb("sharedparam:list", shaped);
            }
            catch (Exception ex)
            {
                SendToWeb("sharedparam:list", new { ok = false, message = ex.Message, items = new List<object>() });
                SendToWeb("revit:error", new { message = ex.Message });
            }
        }

        private void HandleSharedParamStatus(UIApplication app, object payload)
        {
            try
            {
                var status = SharedParameterStatusService.GetStatus(app);
                var shaped = new
                {
                    path = status.Path,
                    isSet = status.IsSet,
                    existsOnDisk = status.ExistsOnDisk,
                    canOpen = status.CanOpen,
                    status = status.Status,
                    statusLabel = status.StatusLabel,
                    warning = status.WarningMessage,
                    errorMessage = status.ErrorMessage
                };
                SendToWeb("sharedparam:status", shaped);
            }
            catch (Exception ex)
            {
                SendToWeb("sharedparam:status", new
                {
                    status = "error",
                    statusLabel = "조회 실패",
                    warning = "Shared Parameter 상태 조회에 실패했습니다.",
                    errorMessage = ex.Message
                });
            }
        }

        private void HandleSharedParamRun(UIApplication app, object payload)
        {
            try
            {
                var sharedStatus = SharedParameterStatusService.GetStatus(app);
                if (sharedStatus == null || !string.Equals(sharedStatus.Status, "ok", StringComparison.OrdinalIgnoreCase))
                {
                    var msg = string.IsNullOrWhiteSpace(sharedStatus?.WarningMessage)
                        ? "Shared Parameter 파일 상태가 올바르지 않습니다."
                        : sharedStatus.WarningMessage;
                    SendToWeb("sharedparam:done", new { ok = false, status = "blocked", message = msg });
                    SendToWeb("paramprop:done", new { ok = false, status = "blocked", message = msg });
                    SendToWeb("revit:error", new { message = "공유 파라미터 연동 실패: " + msg });
                    SendToWeb("sharedparam:status", new
                    {
                        path = sharedStatus?.Path,
                        isSet = sharedStatus?.IsSet,
                        existsOnDisk = sharedStatus?.ExistsOnDisk,
                        canOpen = sharedStatus?.CanOpen,
                        status = sharedStatus?.Status,
                        statusLabel = sharedStatus?.StatusLabel,
                        warning = sharedStatus?.WarningMessage,
                        errorMessage = sharedStatus?.ErrorMessage
                    });
                    return;
                }

                var req = ParamPropagateService.SharedParamRunRequest.FromPayload(payload);
                var res = ParamPropagateService.Run(app, req, ReportParamPropProgress);
                _lastParamResult = res;

                var status = res == null ? ParamPropagateService.RunStatus.Failed : res.Status;
                var ok = status == ParamPropagateService.RunStatus.Succeeded;

                var responsePayload = new Dictionary<string, object>();
                responsePayload["ok"] = ok;
                responsePayload["status"] = status.ToString().ToLowerInvariant();
                responsePayload["message"] = res != null ? res.Message : null;

                if (res != null)
                {
                    var filteredDetails = FilterParamPropIssueDetails(res.Details);
                    responsePayload["report"] = res.Report;
                    responsePayload["details"] = filteredDetails.Select(d => new
                    {
                        kind = d.Kind,
                        family = d.Family,
                        detail = d.Detail
                    }).ToList();
                }

                SendToWeb("sharedparam:done", responsePayload);
                SendToWeb("paramprop:done", responsePayload);

                if (!ok)
                {
                    var msg = res == null ? "실패" : res.Message;
                    SendToWeb("revit:error", new { message = "공유 파라미터 연동 실패: " + msg });
                }
            }
            catch (Exception ex)
            {
                SendToWeb("sharedparam:done", new { ok = false, status = "failed", message = ex.Message });
                SendToWeb("paramprop:done", new { ok = false, status = "failed", message = ex.Message });
                SendToWeb("revit:error", new { message = "공유 파라미터 연동 실패: " + ex.Message });
            }
        }

        private List<ParamPropagateService.SharedParamDetailRow> FilterParamPropIssueDetails(List<ParamPropagateService.SharedParamDetailRow> details)
        {
            var src = details ?? new List<ParamPropagateService.SharedParamDetailRow>();

            var dt = new DataTable("ParamPropUi");
            dt.Columns.Add("Type");
            dt.Columns.Add("Family");
            dt.Columns.Add("Detail");

            foreach (var d in src)
            {
                var row = dt.NewRow();
                row["Type"] = d == null ? "" : (d.Kind ?? "");
                row["Family"] = d == null ? "" : (d.Family ?? "");
                row["Detail"] = d == null ? "" : (d.Detail ?? "");
                dt.Rows.Add(row);
            }

            var filtered = FilterIssueRowsCopy("paramprop", dt);
            var result = new List<ParamPropagateService.SharedParamDetailRow>();
            if (filtered == null) return result;

            foreach (DataRow r in filtered.Rows)
            {
                result.Add(new ParamPropagateService.SharedParamDetailRow
                {
                    Kind = Convert.ToString(r["Type"]),
                    Family = Convert.ToString(r["Family"]),
                    Detail = Convert.ToString(r["Detail"])
                });
            }

            return result;
        }

        private void ReportParamPropProgress(string phase, double phaseProgress, int current, int total, string message, string target)
        {
            SendToWeb("paramprop:progress", new
            {
                phase,
                phaseProgress,
                current,
                total,
                message,
                target
            });
        }

        private void HandleSharedParamExport(UIApplication app, object payload)
        {
            try
            {
                if (_lastParamResult == null)
                {
                    SendToWeb("sharedparam:exported", new { ok = false, message = "최근 실행 결과가 없습니다." });
                    return;
                }

                var doAutoFit = ParseExcelMode(payload);
                var saved = ParamPropagateService.ExportResultToExcel(_lastParamResult, doAutoFit);
                if (string.IsNullOrWhiteSpace(saved))
                {
                    SendToWeb("sharedparam:exported", new { ok = false, message = "엑셀 내보내기가 취소되었습니다." });
                    return;
                }

                SendToWeb("sharedparam:exported", new { ok = true, path = saved });
            }
            catch (Exception ex)
            {
                SendToWeb("sharedparam:exported", new { ok = false, message = ex.Message });
                SendToWeb("revit:error", new { message = "엑셀 내보내기 실패: " + ex.Message });
            }
        }
    }
}
